File: ts/easy/linearSearch.ts
type Structure = {
  id: number;
  name: string;
};

const DATA: Structure[] = [
  {
    id: 1,
    name: 'Jack',
  },
  {
    id: 2,
    name: 'Mag',
  },
  {
    id: 3,
    name: 'Paul',
  },
];

export const linearSearch = (
  data: Structure[],
  targetId: number,
): Structure | null => {
  let i = 0;

  while (i < data.length && data[i].id !== targetId) {
    i++;
  }

  if (data[i]) return data[i];
  return null;
};

console.log('linearSearch', linearSearch(DATA, 2));
console.log('linearSearch', linearSearch(DATA, 10));

export const linearSearch1 = (
  data: Structure[],
  targetId: number,
): Structure | null => {
  const barrier = data.length;
  data[barrier] = { id: barrier, name: '' };
  let i = 0;

  if (targetId > barrier) return null;

  while (data[i].id !== targetId) {
    i++;
  }

  if (i === barrier) return null;
  return data[i];
};

console.log('linearSearch1', linearSearch1(DATA, 2));
console.log('linearSearch1', linearSearch1(DATA, 10));

File: python/easy/linear_search.py
data = [
  {
    "id": 1,
    "name": "Jack",
  },
  {
    "id": 2,
    "name": "Mag",
  },
  {
    "id": 3,
    "name": "Paul",
  },
]

def linear_search(data, target):
  counter = 0

  while counter < len(data) and data[counter]["id"] != target:
    counter += 1
  if target > len(data):
    return None
  return data[counter]
print({ "name": "linear_search", "result": linear_search(data, 2) })
print({ "name": "linear_search", "result": linear_search(data, 10) })
# O(n-k)

def linear_search_1(data, target):
  barrier = len(data)
  data.append({ "id": barrier, "name": "" })
  counter = 0

  if target > barrier:
    return None

  while data[counter]["id"] != target:
    counter += 1

  if counter == barrier:
    return None
  return data[counter]
print({ "name": "linear_search_1", "result": linear_search_1(data, 2) })
print({ "name": "linear_search_1", "result": linear_search_1(data, 10) })
# O(n-k)